### DISCUSSION TITLE
Moving from older working cobbler (2.5 up through 2.8.5) to newer cobbler 3 - questions...

### DISCUSSION THREAD

**Discussion by djtbs1** [cobbler/cobbler#3786] [Q&A]
To anyone who could assist:

Would someone advise me what the major pitfalls / changes are here?
On RHEL 5.x/6.x/7.x my solution was working. The Cobbler I  had on RHEL7.9 was 2.8.5.
I'm moving to RH8.10 so thought I would go to the latest cobbler (3.x) so it would know what RH8.10 is, as a distro.

From a known, working installation fully scripted to a broken one on 3.x because parameter options changed or no longer work. 
For example adding a profile used to be like this:
        cobbler profile add --name=${DISTRO_NAME}-aserver-ng2 \
            --kickstart=/var/lib/cobbler/kickstarts/aserver_rh${rh_specific_version}.ks  --kopts="${KOPTS}"
            
BUT "--kopts" is no longer a valid parameter, it changed to the fully written "--kernel-options".
That's not such a big deal, I went and changed my scripting but now I find out also "--kickstart" is no longer a valid option so what should I do here? .... the documents aren't clear how one would "migrate" this.
my stuff lives in var/lib/cobbler/snippets and var/lib/cobbler/kickstarts, as would be expected. 

IF I should stay on a 2.8.x version someone please let me know and I will revert back if it's known to work without issue on RHEL8.10
PLEASE NOTE any specific recommended versions or additions / file changes that would need done to 2.8.5 for recognizing the RH8.10 official ISO.

I looked at the migrate script but it looks like _a lot_ to go through to see what the script (migrate-data-v2-to-v3) does and what is applicable for my proprietary scripting I already had working.

** EDIT*** - my work laptop was changing the .py script to some HTML marked up with the original python embedded, so the script isn't so large, but still, unsure how much is applicable to my situation.

I've had a class on python, but I'm no expert...
I saw some mention of "autoinstall"- do I change to this parameter from kickstart? Do I move all my ks files to a different directory?
Do I need to do BOTH of those things?

A little clarification specifically for the above command would be helpful.
-or-
as requested above, advice related to staying on 2.8.5 if that's the preferable method.

Best Regards,
Ted

**Comment by SchoolGuy** [comment]
Hey Ted,

sadly with the switch from 2.8.5 to 3.0.0 there was an ownership change and some knowledge was lost there due to the fact that I was a little over my head back then.

I do hope that you have backups from Cobbler 2.8.5 to be able to revert the data directory in case something goes wrong. This is probably the most important thing because fixing inconsistent data is almost impossible since the usual migrations won't apply anymore.

What is your target migration version (3.2.x or 3.3.x)? And from where are you sourcing the RPMs for Cobbler? EPEL?

I do not know how much you have scripted but the most important changes should indeed be:

- ` --kickstart` --> `--autoinstall`
- `--kopts` --> `--kernel-options`
- `--kickstart-meta` --> `--autoinstall-meta`
- The data directory changed from `/var/lib/cobbler/kickstarts/` to `/var/lib/cobbler/autoinstalls`. The snippet directory stayed the same.

To make the script future proof I would recommend using the public XML-RPC API instead of the CLI. At this point in time I sadly am not able to recommend the Terraform Provider that we are having as it is not stable enough. I am making the recommendation of the XML-RPC API as the CLI is currently under redevelopment and will change its interface. The largest advantage is that the script then can be directly executed from a CI pipeline as the communication is purely HTTP-based.

I sadly am not able to offer professional services with the exception if you are a SUSE customer. Then it is possible to open a Packgehub bug that I can then tackle. All other help I am providing here is on a best-effort basis.

I hope that this clears up most of your questions. Feel free to continue inquiring if you need further help.

Regards,
Enno

**Reply by djtbs1** [reply]
Enno-
Thank you for the quick response. I'm using cobbler from the EPEL.
I basically take a set of tarballs over to a rack and unpack them and run my install scripts there on that test rack. So I can basically play with a rack and work through the issues. I have several code branches, so if I have to revert to 2.8.5, it's not the end of the world. 
Once I'm sure my changes work, I put them into my dev area for build into the initial tarballs.

So I will have to try modding my scripts and the locations and assuming that works with the command line changes, hopefully get it up and running. If it doesn't work out, I may hit you up for questions on making RH8.10 work on 2.8.5- I just didn't see an RH8 (el8) version of the 2.8.5....

I'm on a time crunch so don't know that I can do the XML-RPC thing, but may look at it in the future if time permits.

Thanks again!
Best Regards,
Ted

**Reply by SchoolGuy** [reply]
2.8.5 on RHEL 8 will be tricky as not all Python 2 dependencies are available I think. The 2.8.x release series is EOL for sure.

I am having a hard time understanding what you are referring to with tarballs and racks. That is a concept in the context of Cobbler I am not familiar with.

If you are able to share your scripts I may be able to help you a bit more (if they are public).

**Reply by djtbs1** [reply]
Yes, trying to move to python3 off python2 for sure.
I'm old-hat *nix since '95 when linux was a fresh thing to the world so sorry if I'm referring to stuff you may not have used / have access to. :)
racks=  Physical 42U server racks with HPE ProLiant servers (or blade servers) in them and managed network switches. Each server with a specific role that, using cobbler, I automatically install via the kickstart for that role. my code uses the integrated lights out configuration (accessible only on a private network within the rack) to determine the appropriate role for the server. Each rack is basically standalone, not like in a datacenter (though you could put the rack in a datacenter much like any other...)

tarballs= a compilation of files archived into a tar (or compressed tar.gz /.tgz ) for easier distribution. tar came with unix and predates zip and similar.... so I can tar up a bunch of RPMs, repos, ancilliary files into a single package for my field service personnel and when they unpack everything gets put in the appropriate locations. The automation script included in the tarball installs RPMs/repos, etc as necessary. 

SO that way they only have to run 1 command to install, then 1 command that calls cobbler and deploys to the available servers in the rack (it queries the servers to validate what systems can be accessed within the rack).
 
I can't share the scripts, they are proprietary to my workplace- sorry.

Again- thanks for the help.

**Reply by SchoolGuy** [reply]
Don't worry I have enough racks and servers I am responsible for, I am used to those terms. I just wasn't sure how they relate to Cobbler in your setup.

Do I understand you correctly that you nuke the complete Cobbler instance after each install of a rack and then reinstall the RPM and re-bootstrap everything? That sounds like a lot of work that I don't see a benefit in but I am sure that you know best what your infrastructure needs. I am very curious how you do this as I am very interested in making Cobbler as easy to use for as many people as I can.

I expected that your scripts are proprietary but wanted to ask anyway. In case it is of any help, Cobbler has an Ansible inventory plugin that is quite mature to what I have heard (not using it myself though). There is also a Saltstack Formula that I am trying to test at work at the moment. I don't know if that helps but if you are saying that you are from the old guard (that is how I call people at work that are for a long time in the industry) then the chances are high that this is not of help for you.

**Reply by djtbs1** [reply]
Each rack is physically built in our factory area. Bootstrapped from USB on bare metal and all the cobbler bits auto copy onto the system during that process. Other software gets loaded and configured, then the rack will get sent to a customer who may or may not have the rack accessible remotely to us. The rack only gets nuked in event of a need for Disaster recovery. But when I update or re-install our solution, the cobbler bits get re-installed from scratch because sometimes field personnel like to tamper with the config files- so I reinstall every time. Burn me once..... the most dangerous thing is a field person who  starts tampering because they THINK they know how the PXE/DHCP works / is configured, or TFTP, etc.
They see cobbler or dhcp, tftp or other things in ps output and start googling, rather than waiting from a response from support / the developer (me) to resolve the issue. Disabled services or changed service files, etc. MOST times there is either a hardware issue or looking in my logs it can be revealed cables unplugged / moved, etc but the field personnel don't know how to look at mac address lists to track connections, etc.
My cobbler solution looks at the MAC addresses on the system available via the lights out, then when it is going to deploy the role according to the IP address through the process it confirms the MAC address. If the MAC address doesn't match the expected one for that system's connection it doesn't deploy, it stops. That way the server will only get the correct role. I don't know if that method would be useful for anyone else.... but it would also prevent an unexpected system from bootstrapping into your network via PXE (unwanted visitor / MITM attack).
I haven't tried your suggestions yet but expect to later this week, since I'm not in the office, out on leave (technically).

**Reply by djtbs1** [reply]
I've returned from leave and still trying to wrap my head around these changes. 
I've been working through the docs as time permits, but I haven't found the tidbit that would solve my problem thus far.

I checked my dependencies and the only thing I don't have is a generic "gunicorn" rpm- but I do have python3-gunicorn (is a gunicorn rpm besides the python3-gunicorn required??? Reading the description it seems like they do the same thing??) 
But I don't know if that (gunicorn) even has anything to do with the problem I have at this point because PXE is working but not going past the initial efi file.

It used to be (v2.x)  that in /var/lib/tftpboot when cobbler added the system it was added in pxelinux.cfg containing the menuentry at it's mac address. PXE boot loaders were in /var/lib/tftpboot/grub again with mac address entries here. DHCP template points to the boot loaders in grub subdirectory, and chain load occurs.
With Cobbler v3, the loader isn't in grub subfolder it's in the root of tftpboot. If I put grub-x86_64.efi in the location it used to reside cobbler will delete it- so that doesn't work. Changing the DHCP template so it points to the efi file to be in the root of tftp (/var/lib/tftpboot) that will succeed, BUT chain load fails (I just get a grub> prompt) because none of the other MAC address files exist in this same directory. That mac-named file exists in ./grub/system
I thought maybe it had something to do with the bootloaders, but I did a cobbler mkloaders and that didn't help me any that I can tell. 
The logs clearly show PXE trying to get the next file, but the files do not exist in the root of tftpboot area.

I also tried to move from 3.3.3 to 3.3.6 in case one of the known bugs already fixed was the cause of this, but no luck there.

How do I resolve the differences in this new folder structure and file locations to resolve the issue?

**Reply by cinnion** [reply]
@djtbs1 Also working on a similar problem, but only with a partially full 42U rack at home, but gobs of VMs. And been doing UNIX since the 80s (including being a UNIX expert within Bell Labs), so a fellow old-timer. But unlike you, my cobbler instance is running on an ancient Athlon server I have been unable to take out of service for upgrades since CentOS 6, thanks to it also serving DHCP/DNS, Cobbler (probably starting with a 0.4.x version, so I have been through the major version hop a couple of times), LDAP, and worst of all, NFS. But I finally have a working 3.3.6 cobbler server, and am working toward finalizing a Python program which uses the API to dump the configs to a shell script compatible with 3.3.x for final cleanup before executing against my 3.3.6 server running on RL9.  Only problem, the next step is getting PyCharm setup for remote debugging to determine why my script is getting an error. Give me a week or so, and maybe I can help out as well, though I do not yet have EFI booting fully sorted out for my VMs.  (So looking forward to finally getting cobbler/koan working for my RL9 and RHEL9 VM installs instead of hand-rolled kickstarts and shell scripts to kick things off!)

**Reply by SchoolGuy** [reply]
Sorry for the long wait, I am quite busy trying to juggle the project and work with all the related fun that this brings.

So I advise against using any non-tagged version for use. If you develop then of course you want untagged code but the changes during tags are sometimes not migratable to a tagged version and there is no documentation whatsoever of changes from random commit x to tag y. So the current release I would promote for use is 3.3.6.

The current tag 3.3.6 can be installed via RPM and supports Python 3 only. Gunicorn is not a dependency that is needed for 3.3.6, as such you should be alright with installing this on RHEL. I am aware that fence-agents is required and recently that package was split but that should be an easy fix that will be released along with 3.3.7 in the near future.

The folders that you are describing have always been filled by `cobbler sync` and as such all direct edits will be lost. Your workflow should be (according to my understanding of your setup):

- Install Cobbler & required bootloaders via RPM
- Add all Distros, Profiles and Systems (in that order)
- Execute `cobbler mkloaders`
- Execute `cobbler sync`
- Check if everything works and then remove Cobbler from the system again.

The Grub and pxelinux structure that you are describing in the root folder, doesn't match the code and experience that I have with Cobbler. I belive that there is an issue with your setup if that is happening or a bug that I wasn't able to find yet.
